### DISCUSSION TITLE
object not added to Session in future mode

### DISCUSSION THREAD

**Discussion by mehdigmira** [sqlalchemy/sqlalchemy#8172] [Usage Questions]
### Describe the bug

Not sure if this is a bug or a feature, but seems a bit weird to me.
When a session is in future mode, and we have a 1 to many relationship, the _many_ side doesn't seem to be synchronized when we set _one_ side. 
A `session.add` is needed to trigger this synchronization.
In non future mode, the synchronization is automatic.

### To Reproduce

```python
from sqlalchemy import Column, DateTime, ForeignKey, Integer, String, Table, create_engine, func, inspect
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import backref, contains_eager, joinedload, lazyload, relationship, sessionmaker

Base = declarative_base()
engine = create_engine("sqlite://", echo=True, future=True)


class City(Base):
    __tablename__ = "city"
    id = Column(Integer, primary_key=True)
    name = Column(String)

    country_id = Column(Integer, ForeignKey("country.id"))
    country = relationship("Country", back_populates="cities")

    def __repr__(self):
        return self.name


class Country(Base):
    __tablename__ = "country"
    id = Column(Integer, primary_key=True)
    name = Column(String)
    cities = relationship("City", back_populates="country")


Base.metadata.drop_all(engine)
Base.metadata.create_all(engine)
Session = sessionmaker(engine)
session = Session()

session.add(Country(name="france", cities=[City(name="paris")]))

session.commit()

session = Session(future=True)

france = session.query(Country).options(lazyload("*")).one()

marseille = City(name="marseille", country=france)
# session.add(marseille)  # in future mode, assertion breaks  without this

assert len(france.cities) > 1, france.cities  # this passes in non future mode
```

### Error

```
    assert len(france.cities) > 1, france.cities
AssertionError: [paris]
``` 

### Versions

- OS: macos
- Python: 3.8
- SQLAlchemy: 1.4.37
- Database: sqlite
- DBAPI (eg: psycopg, cx_oracle, mysqlclient): sqlite

**Comment by zzzeek** [comment]
hey there -

see https://docs.sqlalchemy.org/en/14/changelog/migration_14.html#cascade-backrefs-behavior-deprecated-for-removal-in-2-0 for background on this.   if you run your program with [RemovedIn20Warnings](https://docs.sqlalchemy.org/en/14/changelog/migration_20.html#migration-to-2-0-step-two-turn-on-removedin20warnings) and run without future=True your program will emit a warning:

> /home/classic/dev/sqlalchemy/test3.py:41: RemovedIn20Warning: "City" object is being merged into a Session along the backref cascade path for relationship "Country.cities"; in SQLAlchemy 2.0, this reverse cascade will not take place.  Set cascade_backrefs to False in either the relationship() or backref() function for the 2.0 behavior; or to set globally for the whole Session, set the future=True flag (Background on this error at: https://sqlalche.me/e/14/s9r1) (Background on SQLAlchemy 2.0 at: https://sqlalche.me/e/b8d9)

**Reply by mehdigmira** [reply]
I've come across this parameter, but this stills seems a bit weird. I think there is bug somewhere:
- if i joinload the relationship before, this works even in future mode 
```python
france = session.query(Country).options(joinedload("*")).one(). # with joinedload this works, if lazyload it will break
marseille = City(name="marseille", country=france)
assert len(france.cities) > 1, france.cities # this passes, cities is synchronized
```
- setting cascade_backrefs=True doesn't change anything to the behavior i describe

**Reply by zzzeek** [reply]
that's because france.cities is now loaded , and creating marseille with country=france appends it to the list of cities, so it's there, in memory. 

however, do session.flush() and you will see:

> `SAWarning: Object of type <City> not in session, add operation along 'Country.cities' will not proceed`

if you session.expire(france) you will also see it failing.

see:

```py
france = session.query(Country).options(joinedload("*")).one() # with joinedload this works, if lazyload it will break

marseille = City(name="marseille", country=france)

assert marseille not in session

# still needed
# session.add(marseille)

session.flush()

session.expire(france)
assert len(france.cities) > 1, france.cities
```
